File: frontend/src/types/goal.ts
export type GoalStatus = "active" | "completed";

// Matches the backend's GoalOut schema exactly - progress figures are
// computed server-side at read time, never stored.
export interface Goal {
  id: number;
  name: string;
  target_amount: number;
  current_amount: number;
  remaining_amount: number;
  progress_percent: number;
  target_date: string; // ISO date string
  status: GoalStatus;
  created_at: string;
  updated_at: string;
}

// Matches backend GoalCreate - current_amount is optional (defaults to 0)
export interface GoalCreateInput {
  name: string;
  target_amount: number;
  target_date: string;
  current_amount?: number;
}

// Matches backend GoalUpdate - deliberately no current_amount; only
// contribute() can move that figure.
export interface GoalUpdateInput {
  name: string;
  target_amount: number;
  target_date: string;
}

File: backend/app/schemas/goal.py
from datetime import date as date_type
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

GoalStatus = Literal["active", "completed"]


class GoalBase(BaseModel):
    name: str = Field(..., min_length=1, max_length=255)
    target_amount: float = Field(..., gt=0, description="Target amount must be greater than zero")
    target_date: date_type


class GoalCreate(GoalBase):
    current_amount: float = Field(default=0.0, ge=0)


class GoalUpdate(GoalBase):
    """
    Deliberately excludes current_amount - contributions are the only way
    to move that figure (POST /goals/{id}/contribute), so editing a goal
    can't silently overwrite progress someone has already logged.
    """

    pass


class GoalContribution(BaseModel):
    amount: float = Field(..., gt=0, description="Contribution must be greater than zero")


class GoalOut(GoalBase):
    model_config = ConfigDict(from_attributes=True)

    id: int
    current_amount: float
    remaining_amount: float
    progress_percent: float
    status: GoalStatus
    created_at: datetime
    updated_at: datetime
